Serialization and range reduction support for Fake Quant/Observer

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26516 Per-channel baseline
* #26520 control of observer/fake-quant operations
* #26622 Default histogram observer
* **#26519 Serialization and range reduction support for Fake Quant/Observer**
* #26876 Fake quantization enhancements for QAT/PTQ support- fix tests

Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

raghuramank10000 committed:
Serialization and range reduction support for Fake Quant/Observer



Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -64,15 +62,38 @@ def extra_repr(self):
+
+        self.quant_min = int(state_dict.pop(prefix + 'quant_min'))
+        self.quant_max = int(state_dict.pop(prefix + 'quant_max'))
+        self.quant_min = bool(state_dict.pop(prefix + 'scale'))
```
scale?

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -64,15 +62,38 @@ def extra_repr(self):
     return partial(fake_quant_cls, **kwargs)
 
 def default_fake_quant(**kwargs):
+    observer = default_observer(reduce_range=True)
```
Dima did some updates to this API I think, could you rebase

jerryzh168 reviewed (changes requested):
see comments

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -65,9 +63,26 @@ def extra_repr(self):
-default_fake_quant = FakeQuantize
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super(FakeQuantize, self)._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + 'quant_min'] = self.quant_min
```
interesting, looks like we have a legit use cases for non-Tensor `register_buffer` (cc @gchanan). No action now, just interesting

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -65,9 +63,26 @@ def extra_repr(self):
+        destination[prefix + 'quant_max'] = self.quant_max
+        destination[prefix + 'scale'] = self.scale
+        destination[prefix + 'zero_point'] = self.zero_point
+        destination[prefix + 'fake_quant_enabled'] = self.fake_quant_enabled
```
is it the state we want to serialize in state_dict? state_dict logically is "learnable parameters", but not really the state of the model. For example we don't serialize `training` attribute today. Why do you need to serialize this runtime state?

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -65,9 +63,26 @@ def extra_repr(self):
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super(FakeQuantize, self)._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + 'quant_min'] = self.quant_min
+        destination[prefix + 'quant_max'] = self.quant_max
```
aren't quant_min/quant_max also parameters of the model? i.e. they are not learnable and one would need to pass them to constructor in code before even loading the statedict?

raghuramank100 replied:
Yes, going by our discussion earlier, these are config params, not learnable params. Will remove them from state dict

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -160,12 +160,28 @@ def _calculate_qparams(self, min_val, max_val):
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super(ObserverBase, self)._save_to_state_dict(destination, prefix, keep_vars)
+
+        destination[prefix + 'dtype'] = self.dtype
```
same comment - state_dict is runtime state serialization, it shouldn't include arguments to constructor

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -264,6 +293,22 @@ def calculate_qparams(self):
 
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super(PerChannelMinMaxObserver, self)._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + 'min_vals'] = self.min_vals
```
can't they just be buffers and get automatic serialization?

raghuramank100 replied:
Yes, for the per-channel case that can be done.

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -13,22 +13,20 @@ class FakeQuantize(Module):
 
-    def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
-                 quant_min=0, quant_max=255):
+    def __init__(self, observer=default_observer, quant_min=0, quant_max=255):
```
with passing observer here it's getting a bit awkward because of nesting. Is FakeQuantize going to be the only base module? Can quant_min/quant_max be deduced from observer's dtype? If so - I'd recommend making FakeQuantize an always on wrapper (handled within prepare_qat) and having qconfig just point to observers instances

raghuramank100 replied:
We want to be able to experiment easily with different choices for quant min/max, say if we want to experiment with lower precisions for weights. In this case, observer may not have all information. Also, fake-quant can potentially learn scale and zero-point via backprop, i.e not have an observer within it. Ideally, we should structure fake-quant to have an abstract base class (like observers). This can be done after we land the basic implementations first.

dzhulgakov replied:
We can do:
```
    def __init__(self, observer=MinMaxObserver, quant_min=0, quant_max=255, **observer_kwargs):
        self.observer = observer(**observer_kwargs)

    QConfig(weight=FakeQuantize.with_args(dtype=np.quint8, quant_min=5))
```

raghuramank100 replied:
Done

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -13,23 +12,20 @@ class FakeQuantize(Module):
         `calcqparam`: A function that calculates quantization parameters
```
the docstring is way out of date - please update

raghuramank100 replied:
Good catch, will update

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -64,6 +60,18 @@ def extra_repr(self):
 
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super(FakeQuantize, self)._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + 'scale'] = self.scale
```
I'd add a comment that we need it because register_buffers doesn't support non-tensors yet (so it'd be easier to fix in the future)

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -234,8 +247,8 @@ class PerChannelMinMaxObserver(_ObserverBase):
         self.ch_axis = ch_axis
-        self.min_vals = None
-        self.max_vals = None
+        self.register_buffer('min_vals', None)
```
maybe just register empty tensor? (torch.tensor([])). Then you don't need to override _load_from_state_dict?

raghuramank100 replied:
Will try that

raghuramank100 replied:
Doesnt work, so sticking with current approach

dzhulgakov reviewed (approved):
Looks good, please address nits before landing

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

raghuramank100 commented:
@pytorchbot Retest this please

raghuramank100 commented:
@pytorchbot retest this please

raghuramank100 commented:
@pytorchbot retest this please

raghuramank10000 committed:
Update on "Serialization and range reduction support for Fake Quant/Observer"


Differential Revision: [D17217408](https://our.internmc.facebook.com/intern/diff/D17217408/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -1,34 +1,67 @@
+        `quant_min`: Tensors are fake-quantized corresponding to the
+        `quant_max`: A function that calculates quantization parameters
         given the stats
+        `observer_kwargs`
```
why do we want this? why not just ask users to apply the observer_kwards with `with_args` on the `observer` argument?

raghuramank100 replied:
Thats because we need to have an observer factory, the kwargs for the observer have to be passed when the observer module is being instantiated.

jerryzh168 replied:
doesn't `with_args` do the same thing?

dzhulgakov replied:
it's a syntactic thing (and was my idea originally). Rationale is that FakeQuant almost "extends" observer. It doesn't prevent doing with_args either, but it's much shorter:

Compare:
```
FakeQuantize.with_args(observer=MinMaxObserver.with_args(dtype=torch.quint8), quant_min=2)
FakeQuantize.with_args(observer=MyCrazyObserver.with_args(dtype=torch.quint8), quant_min=2)
```
with
```
FakeQuantize.with_args(dtype=torch.quint8, quant_min=2)
FakeQuantize.with_args(observer=MyCrazyObserver, dtype=torch.quint8, quant_min=2)
```

jerryzh168 replied:
I see, makes sense

raghuramank100 replied:
Great, was about to write the same thing here: @dzhulgakov : I feel that either API makes sense, and I prefer having explicit with_args for observer. If both of you agree on the current API, we can keep it as is.
@jerryzh168 : Can you stamp this so that I can land this? :-)

jerryzh168 reviewed (approved):
looks good
